### DISCUSSION TITLE
(DONE) Update precession of time to 2020 equations.

### DISCUSSION THREAD

**Discussion by rmathar** [XEphem/XEphem#25] [Ideas]
The equations in precess_hiprec() in precess.c are taken from the Astronomical Almanac 1989 (and the same in the Almanac of 1992) page B18:

```
            zeta_A  = 0.6406161* T + 0.0000839* T*T + 0.0000050* T*T*T;
            z_A     = 0.6406161* T + 0.0003041* T*T + 0.0000051* T*T*T;
            theta_A = 0.5567530* T - 0.0001185* T*T - 0.0000116* T*T*T;
```
where the coefficients are in units of degrees. These are actually used twice, for the forward and backward reduction
to J2000.

In contrast, the equations in the Almanac 2015 (same in 2020) page B52 are 5th order in the centuries T and include constant terms:
```
zeta_A = 2.650545 + T*(2306.083227+T*(0.2988499 + T*(0.01801828
    +T*(-5.971e-6-3.173e-7*T)))) ;
z_A = -2.650545+T*(2306.077181+T*(1.0927348+T*(0.01826837
    +T*(-28.596e-6-2.904e-7*T)))) ;
theta_A = T*(2004.191903-T*(0.4294934+T*(0.04182264+T*(7.089e-6+1.274e-7*T)))) ;
```
where the coefficients are in arcseconds. So an update would be useful supposed one wishes accuracies of the order of 2 arcsec.

**Comment by brandon-rhodes** [comment]
Thanks for the updated equations! These can just be dropped into the source code right where the old ones were?

**Reply by rmathar** [reply]
... and all 3 values must be afterwards divided by 3600 (to convert arcsecs to degrees) before they are used in the sines/cosines, and the comment "From Astronomical Ephermeris 1989, p B18" should be changed to "From Astronomical Almanac 2020, p B52" and the comment from Hayo Hase be dropped... otherwise they are indeed replacements.

**Reply by brandon-rhodes** [reply]
It looks like whoever wants to tackle this update might also want to turn those formulae into a subroutine — am I seeing correctly that currently the exact same code is repeated twice in `precess_hiprec()`?

**Comment by euldulle** [comment]
I am giving a shot at this ; fixed objects are not supposed to be affected ?

**Reply by rmathar** [reply]
Fixed objects (here: defined by those with zero proper motions) _are_ affected, even with observations at the 2000 epoch, because the new equations have constant terms of the order of 2 arcseconds and the old equations start with terms linear in T. This is why my earlier comment talks about "accuracies of the order of 2 arcsec."

**Reply by euldulle** [reply]
Ok, my bad, I was trying to mass probe the results, but I am probably probing at the wrong place.

**Reply by euldulle** [reply]
I must be missing something big here : 
I see correct results out of hires_precess(), but somehow these results 
do not make their way to the skyview interface, that behaves as if precession was never applied.
Vega 2000.0 shows RA: 18h36m56.34s DE:+38°47'01.03"
Vega 2100.0 should show something like RA: 18h40m19.60s DE:+38°53'06.23"
but instead I see 18h36m58.06 and 38°47'30 which suggest only proper motion has been applied, not precession.
What am I missing ?

**Reply by brandon-rhodes** [reply]
@euldulle — In the main panel, you might need to click on "Equinox" then click the button "Of Date".

**Reply by euldulle** [reply]
Ok, no comment...
At least I was right diagnosing I was missing an elephant :)
Thanks.

**Reply by euldulle** [reply]
It took me some time to figure what is going on here (forget the elephant...), as I was unable
to have these 2 arcsec offsets pop up in the interface.

The constant offsets are misleading, they do not translate as is in the final results,
so at T=0, z_A has an offset in the new formulae but after the rotation, there is virtually
no difference between both sets of formulae, at least nowhere near the arcsecond level.  
No coincidence, the new set of formulae is designed for that. And its quite impressive to see :
at T=0, the results from the rotation match the constant term at an incredible level.

Sample :
```
a_in 230.281712979347 dec_in  -5.824852776449 from 2000.0000 to 2000.0014
bwd a_out 230.281712979347 dec_out  -5.824852776449 z_A  -0.000736262500 ATAN 230.282449241847
```
which could read as follows : 
enter precess with alpha_in = 230.281712979347, the second line is the backward jump from 
2000.000 to 2000.0014 (which is in fact a forward jump, the epoch of the star coordinates 
being a bit after 2000.0) ; compute the rotation, you will find 230.282449241847 ;
add the constant term z_A=-0.000736262500 (=-2.650545 arcsec) and you get exactly the
same alpha that you entered with... Baffling.

I delved a bit in the paper https://www.aanda.org/articles/aa/pdf/2003/48/aa4068.pdf
and we are in a different world.
Even the old formulae are probably well within 1 arcsec for centuries. 
The accuracy of the new formulae are in the microarcsec world.

Quote :
> The following series with a 0.1microarcsec level of precision matches the canonical 
> 4-rotation series to sub-microarcsecond accuracy over 4 centuries...

Should be enough for us mere mortals :)
